// @redclover/koru-sdk/dist/index.mjs
/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

function __rest(s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

/**
 * Abstract base class for building Koru widgets.
 * Handles authorization, caching, lifecycle management, and provides helper utilities.
 *
 * @example
 * ```typescript
 * class MyWidget extends KoruWidget {
 *   constructor() {
 *     super({ name: 'my-widget', version: '1.0.0' });
 *   }
 *
 *   async onInit(config) {
 *     // Initialize widget state
 *   }
 *
 *   async onRender(config) {
 *     // Render widget UI
 *     this.container = this.createElement('div', {
 *       className: 'my-widget',
 *       children: ['Hello World']
 *     });
 *     document.body.appendChild(this.container);
 *   }
 *
 *   async onDestroy() {
 *     // Cleanup
 *     this.container?.remove();
 *   }
 * }
 *
 * new MyWidget().start();
 * ```
 */
class KoruWidget {
    /**
     * Creates a new widget instance.
     * Extracts configuration from script tag data attributes.
     *
     * @param widgetOptions - Widget configuration including name, version, and options
     * @throws {Error} If script tag or required data attributes are missing
     *
     * @example
     * ```typescript
     * constructor() {
     *   super({
     *     name: 'my-widget',
     *     version: '1.0.0',
     *     options: {
     *       cache: true,
     *       debug: true,
     *       analytics: false
     *     }
     *   });
     * }
     * ```
     */
    constructor(widgetOptions) {
        /** Current widget configuration from Koru */
        this.config = {};
        /** Full authorization response including app and website metadata */
        this.authData = null;
        /** Main container element for the widget (typically set in onRender) */
        this.container = null;
        this.isInitialized = false;
        this.widgetName = widgetOptions.name;
        this.widgetVersion = widgetOptions.version;
        // Default options
        this.options = Object.assign({ cache: true, cacheDuration: 3600, retryAttempts: 3, retryDelay: 1000, analytics: false, debug: false }, widgetOptions.options);
        // Extract params from script tag
        const script = this.getCurrentScript();
        if (!script) {
            throw new Error('[Widget SDK] Could not find script tag');
        }
        this.websiteId = script.getAttribute('data-website-id') || '';
        this.appId = script.getAttribute('data-app-id') || '';
        this.koruUrl = script.getAttribute('data-app-manager-url') || '';
        this.customData = script.getAttribute('data-custom-data') || '';
        if (!this.websiteId || !this.appId || !this.koruUrl) {
            throw new Error('[Widget SDK] Missing required data attributes');
        }
        this.log('SDK initialized', { websiteId: this.websiteId, appId: this.appId });
    }
    /**
     * Starts the widget lifecycle.
     * Waits for DOM, authorizes with Koru, then calls onInit and onRender hooks.
     *
     * @returns Promise that resolves when widget is fully initialized
     * @throws {Error} If authorization fails or lifecycle hooks throw errors
     *
     * @example
     * ```typescript
     * const widget = new MyWidget();
     * await widget.start();
     * ```
     */
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.waitForDOM();
                this.authData = yield this.authorize();
                if (!this.authData.authorized) {
                    this.log('Widget not authorized');
                    return;
                }
                this.config = this.authData.config;
                yield this.onInit(this.config);
                yield this.onRender(this.config);
                this.isInitialized = true;
                this.log('Widget started successfully');
                if (this.options.analytics) {
                    this.track('widget_loaded', {
                        widget: this.widgetName,
                        version: this.widgetVersion
                    });
                }
            }
            catch (error) {
                this.handleError('Failed to start widget', error);
            }
        });
    }
    /**
     * Stops the widget and performs cleanup.
     * Calls the onDestroy hook to remove DOM elements and event listeners.
     *
     * @returns Promise that resolves when cleanup is complete
     *
     * @example
     * ```typescript
     * await widget.stop();
     * ```
     */
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.onDestroy();
                this.isInitialized = false;
                this.log('Widget stopped');
            }
            catch (error) {
                this.handleError('Failed to stop widget', error);
            }
        });
    }
    /**
     * Reloads the widget with fresh configuration from Koru.
     * Clears cache and re-authorizes. Calls onConfigUpdate if implemented, otherwise re-renders.
     *
     * @returns Promise that resolves when reload is complete
     *
     * @example
     * ```typescript
     * // Reload widget when configuration changes
     * await widget.reload();
     * ```
     */
    reload() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                this.clearCache();
                this.authData = yield this.authorize();
                if (!this.authData.authorized)
                    return;
                const newConfig = this.authData.config;
                if (this.onConfigUpdate) {
                    yield this.onConfigUpdate(newConfig);
                }
                else {
                    yield this.onDestroy();
                    yield this.onRender(newConfig);
                }
                this.config = newConfig;
                this.log('Widget reloaded');
            }
            catch (error) {
                this.handleError('Failed to reload widget', error);
            }
        });
    }
    // ==================== AUTHORIZATION ====================
    authorize() {
        return __awaiter(this, void 0, void 0, function* () {
            // Check for Koru preview mode first
            if (typeof window !== 'undefined' && window.__KORU_PREVIEW_CONFIG__) {
                this.log('Using Koru preview config');
                return {
                    authorized: true,
                    config: window.__KORU_PREVIEW_CONFIG__,
                    token: 'preview-mode',
                    app: {
                        id: this.appId,
                        name: 'Preview Mode',
                        description: 'Widget running in Koru preview mode'
                    },
                    website: {
                        id: this.websiteId,
                        url: window.location.origin,
                        is_ecommerce: false,
                        customer: 'preview'
                    }
                };
            }
            if (this.options.cache) {
                const cached = this.getFromCache();
                if (cached) {
                    this.log('Using cached authorization');
                    return cached;
                }
            }
            let lastError = null;
            for (let attempt = 1; attempt <= this.options.retryAttempts; attempt++) {
                try {
                    this.log(`Authorization attempt ${attempt}/${this.options.retryAttempts}`);
                    let url = `${this.koruUrl}/api/auth/widget?website_id=${this.websiteId}&app_id=${this.appId}`;
                    if (this.customData) {
                        url += `&custom_data=${encodeURIComponent(this.customData)}`;
                    }
                    const response = yield fetch(url, {
                        method: 'GET',
                        headers: { 'Content-Type': 'application/json' },
                    });
                    if (!response.ok) {
                        throw new Error(`Authorization failed: ${response.status}`);
                    }
                    const data = yield response.json();
                    if (this.options.cache) {
                        this.saveToCache(data);
                    }
                    return data;
                }
                catch (error) {
                    lastError = error;
                    this.log(`Authorization attempt ${attempt} failed:`, error);
                    if (attempt < this.options.retryAttempts) {
                        yield this.sleep(this.options.retryDelay);
                    }
                }
            }
            throw lastError || new Error('Authorization failed');
        });
    }
    // ==================== HELPER METHODS ====================
    /**
     * Creates a DOM element with the specified tag and properties.
     * Provides a convenient way to build UI elements with type safety.
     *
     * @template K - HTML element tag name
     * @param tag - HTML tag name (e.g., 'div', 'button', 'span')
     * @param props - Element properties including className, style, onClick, and children
     * @returns The created HTML element
     *
     * @example
     * ```typescript
     * // Create a button with click handler
     * const button = this.createElement('button', {
     *   className: 'btn btn-primary',
     *   style: { padding: '10px', backgroundColor: 'blue' },
     *   onClick: () => console.log('clicked'),
     *   children: ['Click Me']
     * });
     *
     * // Create nested elements
     * const card = this.createElement('div', {
     *   className: 'card',
     *   children: [
     *     this.createElement('h2', { children: ['Title'] }),
     *     this.createElement('p', { children: ['Description'] })
     *   ]
     * });
     * ```
     */
    createElement(tag, props) {
        const element = document.createElement(tag);
        if (props) {
            const { className, style, onClick, children } = props, rest = __rest(props, ["className", "style", "onClick", "children"]);
            if (className)
                element.className = className;
            if (style)
                Object.assign(element.style, style);
            if (onClick)
                element.addEventListener('click', onClick);
            if (children) {
                children.forEach(child => {
                    if (typeof child === 'string') {
                        element.appendChild(document.createTextNode(child));
                    }
                    else {
                        element.appendChild(child);
                    }
                });
            }
            Object.assign(element, rest);
        }
        return element;
    }
    /**
     * Detects if the current device is a mobile device.
     * Checks user agent string for common mobile device identifiers.
     *
     * @returns true if mobile device, false otherwise
     *
     * @example
     * ```typescript
     * async onRender(config) {
     *   const layout = this.isMobile() ? 'mobile' : 'desktop';
     *   this.container = this.createElement('div', {
     *     className: `widget-${layout}`
     *   });
     * }
     * ```
     */
    isMobile() {
        return /Android|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator.userAgent);
    }
    /**
     * Tracks an analytics event to Koru.
     * Only sends events if analytics option is enabled.
     *
     * @param eventName - Name of the event to track
     * @param eventData - Optional event metadata
     *
     * @example
     * ```typescript
     * // Track button click
     * this.track('button_clicked', {
     *   button_id: 'cta',
     *   timestamp: Date.now()
     * });
     *
     * // Track page view
     * this.track('page_viewed', {
     *   page: 'home'
     * });
     * ```
     */
    track(eventName, eventData) {
        if (!this.options.analytics)
            return;
        try {
            const url = `${this.koruUrl}/api/analytics`;
            fetch(url, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    website_id: this.websiteId,
                    app_id: this.appId,
                    event_type: eventName,
                    event_data: Object.assign({ widget: this.widgetName, version: this.widgetVersion, custom_data: this.customData }, eventData)
                })
            }).catch(err => this.log('Analytics error:', err));
        }
        catch (error) {
            this.log('Failed to track event:', error);
        }
    }
    /**
     * Logs a debug message to the console.
     * Only logs if debug option is enabled.
     *
     * @param message - Message to log
     * @param args - Additional arguments to log
     *
     * @example
     * ```typescript
     * this.log('Widget initialized', { config: this.config });
     * this.log('Fetching data from', apiUrl);
     * ```
     */
    log(message, ...args) {
        if (this.options.debug) {
            console.log(`[${this.widgetName}]`, message, ...args);
        }
    }
    handleError(message, error) {
        console.error(`[${this.widgetName}] ${message}:`, error);
        if (this.options.analytics) {
            this.track('widget_error', {
                error: error instanceof Error ? error.message : String(error),
                message
            });
        }
    }
    // ==================== PRIVATE HELPERS ====================
    getCurrentScript() {
        return document.currentScript;
    }
    waitForDOM() {
        return __awaiter(this, void 0, void 0, function* () {
            if (document.readyState === 'loading') {
                return new Promise(resolve => {
                    document.addEventListener('DOMContentLoaded', () => resolve());
                });
            }
        });
    }
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    getCacheKey() {
        return `koru_widget_${this.websiteId}_${this.appId}`;
    }
    getFromCache() {
        try {
            const key = this.getCacheKey();
            const cached = localStorage.getItem(key);
            if (!cached)
                return null;
            const { data, timestamp } = JSON.parse(cached);
            const age = (Date.now() - timestamp) / 1000;
            if (age > this.options.cacheDuration) {
                this.clearCache();
                return null;
            }
            return data;
        }
        catch (_a) {
            return null;
        }
    }
    saveToCache(data) {
        try {
            const key = this.getCacheKey();
            localStorage.setItem(key, JSON.stringify({
                data,
                timestamp: Date.now()
            }));
        }
        catch (error) {
            this.log('Failed to cache auth data:', error);
        }
    }
    clearCache() {
        try {
            const key = this.getCacheKey();
            localStorage.removeItem(key);
        }
        catch (_a) {
            // Ignore errors
        }
    }
}

export { KoruWidget };
